lib/propolis/src/hw/virtio/softnpu.rs

```diff
@@ -0,0 +1,1061 @@
+            data_handles.clone(),
+            virtio.clone(),
+            pipeline.clone(),
+            log.clone(),
+        );
+
+        Ok(Arc::new(SoftNpu {
+            data_links,
+            data_handles,
+            virtio,
+            pci_port,
+            uart,
+            p9fs,
+            pipeline,
+            booted: Mutex::new(false),
+            log,
+        }))
+    }
+
+    /// Generate a mac address with the Oxide OUI for the leading bits and then
+    /// something random in the range of 0xf00000 - 0xf00000 per RFD 174.
+    fn generate_mac() -> [u8; 6] {
+        let mut rng = rand::thread_rng();
+        let m = rng.gen_range::<u32, _>(0xf00000..0xffffff).to_le_bytes();
+        [0xa8, 0x40, 0x25, m[0], m[1], m[2]]
+    }
+
+    /// Set up a dlpi handle for each data link.
+    fn data_handles(data_links: &Vec<String>) -> Result<Vec<dlpi::DlpiHandle>> {
+        let mut handles = Vec::new();
+        for x in data_links {
+            let h = dlpi::open(x, dlpi::sys::DLPI_RAW)?;
+
+            // Although we bind to the IPv6 SAP (Ethertype), the DL_PROMISC_SAP
+            // allows us to pick up everything. Binding to *something* to start
+            // with appears to be required to get packets.
+            dlpi::bind(h, 0x86dd)?;
+            dlpi::promisc_on(h, dlpi::sys::DL_PROMISC_MULTI)?;
+            dlpi::promisc_on(h, dlpi::sys::DL_PROMISC_SAP)?;
+            dlpi::promisc_on(h, dlpi::sys::DL_PROMISC_PHYS)?;
+            dlpi::promisc_on(h, dlpi::sys::DL_PROMISC_RX_ONLY)?;
+            handles.push(h);
+        }
+        Ok(handles)
+    }
+
+    /// Start the management handler for servicing requests from the guest over
+    /// the provided uart device.
+    fn run_management_handler_thread(&self) {
+        info!(self.log, "softnpu: running management handler");
+        self.uart.set_autodiscard(false);
+
+        let log = self.log.clone();
+        let uart = self.uart.clone();
+        let pipeline = self.pipeline.clone();
+        let radix = self.data_links.len();
+
+        spawn(move || Self::management_handler(uart, pipeline, radix, log));
+    }
+
+    fn management_handler(
+        uart: Arc<LpcUart>,
+        pipeline: Arc<Mutex<Option<(Library, Box<dyn Pipeline>)>>>,
+        radix: usize,
+        log: Logger,
+    ) {
+        info!(log, "management handler thread started");
+        loop {
+            let r = ManagementMessageReader::new(uart.clone(), log.clone());
+            let msg = r.read();
+            info!(log, "received management message: {:#?}", msg);
+
+            let pipeline = pipeline.clone();
+            let uart = uart.clone();
+            let log = log.clone();
+            handle_management_message(msg, pipeline, uart, radix, log.clone());
+            info!(log, "handled management message");
+        }
+    }
+}
+
+impl Entity for SoftNpu {
+    fn type_name(&self) -> &'static str {
+        "softnpu"
+    }
+
+    fn start(&self) {
```

Comment:
Given that only the `start` handler is implemented, is it safe to say that falcon guests (or at least those using SoftNPU) are never expected to reboot? (Or for that matter, cleanly halt)

Reply:
An interesting question. For the primary thing that SoftNPU currently emulates, a sidecar switch - the sidecar will indeed keep on going if the attached scrimlet reboots, halts, etc. So in that sense, I think we have the desired behavior here, where the guest (scrimlet) reboots and comes back to find the sidecar not having gone down with it.

If we start using SoftNpu for emulating something like a programmable NIC, that would be a different story.

Reply:
To be clear: I'm not suggesting that we should reset the emulated switch state, but as-is, we're ignoring a bunch of expected state management expected of an emulated (PCI) device if the guest undergoes reboot.

If you really do not expect a falcon guest (at least one with a SoftNPU device attached) to undergo reboot, then this is probably fine, but as-is, it is breaking some of the rules/expectations around how device emulation acts when the instance is walking through expected state transitions for a reboot or halt.

Reply:
I'd overlooked the fact that the `PciVirtioSoftNpuPort` device had wired up the appropriate events for ensuring a PciVirtio reset when the guest itself is rebooted.  While there are some improvements that could be made around how the dlpi recv thread(s) are being managed (they'll likely prevent clean propolis exit on instance halt), that is something which could be addressed in a follow-up.  Clearly the tools in place for using Falcon today already take that into account.

Reply:
I've added an issue to track this 
- https://github.com/oxidecomputer/propolis/issues/261